SCSI persistent reservation with device plugin

alicefr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
SCSI protocol offers dedicated commands in order to reserve and control access to the LUNs. This can be used to prevent data corruption if the disk is shared by multiple VMs (or more in general processes).

The SCSI persistent reservation is handled by the [qemu-pr-helper](https://qemu.readthedocs.io/en/latest/tools/qemu-pr-helper.html). The pr-helper is a privileged daemon that can be either started by libvirt directly or managed externally. 

In case of KubeVirt, the qemu-pr-helper needs to be started externally because it requires high privileges in order to perform the persistent SCSI reservation. Afterward, the pr-helper socket is accessed by the unprivileged virt-launcher pod for enabling the SCSI persistent reservation. 

This PR is a second version of https://github.com/kubevirt/kubevirt/pull/8210 using device plugin framework. 
The pr-helper daemon is deployed in the same pod as virt-handler in a separate container. This PR introduces a new device plugin for mounting the pr-socket inside the virt-launcher container if it requests the resource `devices.kubevirt.io/pr-helper` and is enabled by the `reservations` field in the VMI declaration.

VMI example:
```yaml
    devices:
      disks:
      - name: mypvcdisk
        lun:
          reservations: true
```

The device plugin framework doesn't offer any access control. However, having access to the pr-helper socket isn't enough to perform the reservation, as the pod requires also access to the SCSI device. The SCSI device is managed through PVCs and k8s RBAC.

This feature is controlled by the feature gate `PersistentReservation`:
```yaml
  configuration:
    developerConfiguration:
      featureGates:
      -  PersistentReservation
```

Once the feature gate is enabled, then the additional container with the `qemu-pr-helper` is deployed inside the virt-handler pod. Enabling (or removing) the feature gate causes the redeployment of the virt-handler pod.

An important aspect of this feature is that the SCSI persistent reservation doesn't support migration. Even if you apply the reservation to an RWX PVC provisioning SCSI devices, the restriction is due to the reservation done by the initiator on the node. The VM could be migrated but not the reservation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8115

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Adding SCSI persistent reservation
```

alicefr commented:
/cc @vladikr this is the scsi persistent reservation version with the device plugin

alicefr commented:
@vladikr @xpivarc it is a question for you.

About the feature gate, I'm not sure if we should simply avoid starting the device plugin for the socket or also deploying the pr-helper container.
Currently, the pr-helper is deployed as an additional container inside the virt-handler pod. If we deploy the pr-helper container based on the feature gate, then if it was disabled and we enable it in a second moment, the container isn't automatically deployed if we don't restart the virt-handler pod entirely. Therefore, the simplest solution would be to deploy the pr-helper container in all cases and simply enable or not the device plugin.

Another alternative could be to deploy the pr-helper in a separate daemon set. I have avoid so fare this solution because it complicates the deployment of KubeVirt as we need an additional daemonset to manage.

vasiliy-ul reviewed (commented):
> Therefore, the simplest solution would be to deploy the pr-helper container in all cases and simply enable or not the device plugin.

Since the `pr-helper` container is privileged, it brings additional security risks. I would try to avoid running it at all if it is not explicitly enabled by the end-user via the feature gate.

> Another alternative could be to deploy the pr-helper in a separate daemon set. I have avoid so fare this solution because it complicates the deployment of KubeVirt as we need an additional daemonset to manage.

I think we need to consider how critical the impact of restarting virt-handler might be. As I understand, it should not affect running workloads (hopefully :P ). If it's smth we can live with, then we probably just need to properly document the behavior (i.e. that virt-handler needs to be restarted if the feature gate is enabled/disabled).

alicefr commented on tests/storage/scsi.go:
```diff
@@ -0,0 +1,184 @@
+	"kubevirt.io/kubevirt/tests/util"
+)
+
+var _ = SIGDescribe("[Serial][sig-storage]SCSI persistent reservation", Serial, func() {
```
I'm updating the kubevirt cr to temporarily enable the PersistentReservation feature gate. It was complaining that the test cannot run in parallel if it modifies a cluster-shared object. Maybe here, I'm misunderstanding how the feature gates work with the tests.

vasiliy-ul replied:
Probably the suggestion from the next comment can solve that https://github.com/kubevirt/kubevirt/pull/9177#discussion_r1100074433

alicefr replied:
yup, it solved it

alicefr commented on tests/storage/scsi.go:
```diff
@@ -0,0 +1,184 @@
+		// TODO: remove this check once we have built the targetcli image for arm64 or fixed the bazel issue
+		checks.SkipIfARM64(testsuite.Arch, "targetcli test images not currently built for arm64")
+
+		if !checks.HasFeature(virtconfig.PersistentReservation) {
```
oh yes, I didn't know this, thanks!

alicefr commented:
> > Therefore, the simplest solution would be to deploy the pr-helper container in all cases and simply enable or not the device plugin.
> 
> Since the `pr-helper` container is privileged, it brings additional security risks. I would try to avoid running it at all if it is not explicitly enabled by the end-user via the feature gate.
> 

Yes, I'd like to have also some more dynamic mechanism.

> > Another alternative could be to deploy the pr-helper in a separate daemon set. I have avoid so fare this solution because it complicates the deployment of KubeVirt as we need an additional daemonset to manage.
> 
> I think we need to consider how critical the impact of restarting virt-handler might be. As I understand, it should not affect running workloads (hopefully :P ). If it's smth we can live with, then we probably just need to properly document the behavior (i.e. that virt-handler needs to be restarted if the feature gate is enabled/disabled).

Again, the other option could be to introduce new daemonset just for the pr-helper. Virt-controller could create (or remove) this new damonset based on the feature gates. The part that worries me more is the upgrade path. It is just an additional components to upgrade. 
I'd really appreciate on this feedback from @xpivarc and @vladikr :)

alicefr commented on tests/storage/scsi.go:
```diff
@@ -0,0 +1,184 @@
+				&expect.BExp{R: "  LIO-ORG   disk0             4.0 \r\n  Peripheral device type: disk\r\n"},
+			}, 30)).To(Succeed())
+
+			Eventually(checkPersistentReservation(vmi), 10*time.Second, 60*time.Second).Should(BeTrue())
```
Good catch thanks!

alicefr commented on tests/storage/scsi.go:
```diff
@@ -0,0 +1,184 @@
+			tests.EnableFeatureGate(virtconfig.PersistentReservation)
+			disableFG = true
+		}
+		Expect(err).ToNot(HaveOccurred())
```
Needed to be moved up, it is fixed since I removed the fg, it is back to its place :)

alicefr commented on pkg/virt-operator/resource/generate/components/daemonsets.go:
```diff
@@ -20,9 +22,34 @@ import (
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: pointer.Bool(true),
+		},
+		Stdin: true,
```
Removed it

alicefr commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -399,6 +399,11 @@ func (app *virtHandlerApp) Run() {
 	go gracefulShutdownInformer.Run(stop)
 	go domainSharedInformer.Run(stop)
 
+	prSockDir, err := safepath.JoinAndResolveWithRelativeRoot("/", "/var/run/kubevirt/daemons/pr")
```
Yes, good idea

alicefr commented:
I added the ability to deploy the pr-helper container if the feature gate is set or not. This doesn't require any manual operations and the virt-handler pod will be automatically redeployed

mhenriks commented:
Looks great, one high level question for now

```
devices.kubevirt.io/kvm: 1k
devices.kubevirt.io/pr-helper: "1"
devices.kubevirt.io/tun: 1k
devices.kubevirt.io/vhost-net: 1k
```

Can pr-helper be shared or is there only one connection per node allowed?

alicefr commented:
> Looks great, one high level question for now
> 
> ```
> devices.kubevirt.io/kvm: 1k
> devices.kubevirt.io/pr-helper: "1"
> devices.kubevirt.io/tun: 1k
> devices.kubevirt.io/vhost-net: 1k
> ```
> 
> Can pr-helper be shared or is there only one connection per node allowed?

We will have a pr-helper per node and all the VMs on that node will connect to it. 

Generally, you could have also a pr-helper-pr VM. This is the model used when libvirt deploys directly the pr-helper. However, the pr-helper requires high privileges and therefore we cannot apply this model.
The daemon can serve multiple VMs, for example this is the model used when you start the pr-helper as a systemd service

alicefr commented:
/retest-required

mhenriks commented:
> > Looks great, one high level question for now
> > ```
> > devices.kubevirt.io/kvm: 1k
> > devices.kubevirt.io/pr-helper: "1"
> > devices.kubevirt.io/tun: 1k
> > devices.kubevirt.io/vhost-net: 1k
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can pr-helper be shared or is there only one connection per node allowed?
> 
> We will have a pr-helper per node and all the VMs requesting the persistent reservation on that node will connect to it.
> 
> Generally, you could have also a pr-helper-pr VM. This is the model used when libvirt deploys directly the pr-helper. However, the pr-helper requires high privileges and therefore we cannot apply this model. The daemon can serve multiple VMs, for example this is the model used when you start the pr-helper as a systemd service

My question was regarding "1" vs 1K reported for the other devices we make available.

alicefr commented:
> > > Looks great, one high level question for now
> > > ```
> > > devices.kubevirt.io/kvm: 1k
> > > devices.kubevirt.io/pr-helper: "1"
> > > devices.kubevirt.io/tun: 1k
> > > devices.kubevirt.io/vhost-net: 1k
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Can pr-helper be shared or is there only one connection per node allowed?
> > 
> > 
> > We will have a pr-helper per node and all the VMs requesting the persistent reservation on that node will connect to it.
> > Generally, you could have also a pr-helper-pr VM. This is the model used when libvirt deploys directly the pr-helper. However, the pr-helper requires high privileges and therefore we cannot apply this model. The daemon can serve multiple VMs, for example this is the model used when you start the pr-helper as a systemd service
> 
> My question was regarding "1" vs 1K reported for the other devices we make available. But maybe I'm misunderstanding how devices are made available. My assumption is that this would allow one pod to use the device on the node

Nice catch! Yes, it was missing the set of the maxDevices (now added by https://github.com/kubevirt/kubevirt/pull/9177/commits/ee7c5daff4a35d77bbe23dc033526b633f56de15).
I have also added a new test with multiple VMs accessing the same LUN and requesting the persistent reservation. This should cover also this case.

mhenriks commented:
/retest-required

vasiliy-ul commented on tests/utils.go:
```diff
@@ -2284,9 +2284,95 @@ func GetDefaultVirtControllerDeployment(namespace string, config *util.KubeVirtD
+					Name: dbus,
+					VolumeSource: k8sv1.VolumeSource{
+						HostPath: &k8sv1.HostPathVolumeSource{
+							Path: "/var/run/dbus",
```
Probably some stupid questions from a person not very familiar with targetcli and how it works :)

1. Do I understand correctly that here `targetcli` will talk to a deamon running on the **node** through the system dbus socket?
2. Why does a cli tool need `/sys/kernel/config` and `/lib/modules`?

alicefr replied:
`targetcli` needs to load the kernel module `tcm_loop` and mount the configs under /sys/kernel/config/

alicefr replied:
Actually, I'm not sure if it is targetcli that is doing it automatically or kubevirtci has already those configured, but those are pre-steps. I just realized I don't have in the test :/ I need to check if I need to check the environment and eventually execute them. I have created a simple [gist](https://gist.github.com/alicefr/c2e4221d7c8834a2b8746d510692d86c) to list the steps for the device creation

alicefr replied:
ok, I just checked and `tcm_loop` isn't automatically loaded in kubevirtci, so it needs to be targetcli that does it

vasiliy-ul replied:
> modprobe tcm_loop

Isn't the module loaded automatically when targetcli tries to use it?

> mount -t configfs configfs /sys/kernel/config/

I think for that you do not need to use HostPath. I think you can just create a directory and mount configs in the container. Or maybe use an EmptyDir instead? 


Still, why does it need dbus from the node?

alicefr replied:
Even if it is targetcli that automatically loads the module, it needs to have access to the module to load.
I can try to add empty dir for mounting configs, thanks for the suggestion!
When I was playing with targetcli, it was requiring having access to the socket. It was at the begging, I'll double-check again if we can avoid it. Just targetcli is a generic utility for dealing with SCSI disks and it still might requiring for other operations

alicefr replied:
I have double-checked (using the setup available[here](https://github.com/alicefr/targetcli-containers/blob/main/start-container.sh)) without the path for configs seems working, however without access to dbus, I get an error: `g-io-error-quark: Could not connect: No such file or directory (1)
`

alicefr commented:
/test-required

alicefr commented:
/retest

alicefr commented:
/retest-required

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-compute

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-storage

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.25-sig-compute-migrations

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-network

alicefr commented:
[One](https://github.com/kubevirt/kubevirt/blob/main/tests/operator/operator.go#L1599) of the functional tests in the operator series tries to install the image from the previous version, and of course we don't have the pr-helper image. What will be a proper solution? Detect the latest kubevirt version and unset the feature gate `PersistentReservation` if it is the latest version that doesn't support persistent reservation? Or does anyone have a simpler solution?

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute-root

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-storage

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-network

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute

alicefr commented:
> [One](https://github.com/kubevirt/kubevirt/blob/main/tests/operator/operator.go#L1599) of the functional tests in the operator series tries to install the image from the previous version, and of course we don't have the pr-helper image. What will be a proper solution? Detect the latest kubevirt version and unset the feature gate `PersistentReservation` if it is the latest version that doesn't support persistent reservation. Or does anyone have a simpler solution? @mhenriks @vasiliy-ul @xpivarc any idea?
> 
> UPDATE: Example at: https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/9177/pull-kubevirt-e2e-k8s-1.24-operator/1630499694272581632/artifacts/k8s-reporter/1_pods.log
> 
> ```json
>  "name": "pr-helper",
>   "state": {
>      "waiting": {
>               "reason": "ImagePullBackOff",
>               "message": "Back-off pulling image \"quay.io/kubevirt/pr-helper:v0.58.1\""
>             }
> ```

Proposed solution: https://github.com/kubevirt/kubevirt/pull/9177/commits/fe175d12cde8cafecdb634fbc90e6210ba06f382

alicefr commented:
There are a lot of failures on this PR that are due to timeout, this particularly affects `sig-compute-*` and `sig-network` . Sometimes there the containers in virt-handler are in `context deadline exceeded` or the virt-launcher pods are simply very slow to be deployed (like for the test [[`sig-network] [ref_id:1182]Probes for liveness should not fail the VMI
[test_id:1201][posneg:positive]with working HTTP probe and http server on
ipv6`](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/9177/pull-kubevirt-e2e-k8s-1.24-sig-network/1630869077729742848)
where the [virt-launcher pod virt-launcher-testvmi-cjvcs-h6x7f  is
initializing](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/9177/pull-kubevirt-e2e-k8s-1.24-sig-network/1630869077729742848/artifacts/k8s-reporter-parallel/3/1_pods.log))).

I'm not sure why this is happening but I have a suspicion that is due to the fact that enabling the `PersistentReservation` triggers virt-handler to be redeployed and this slows the entire system down. Should we enable the feature
gate only for the `storage` and `operator` lanes (imo, it will be good also to test the upgrade path)? In this way, we could reduce the number of virt-handler restarts. In a production environment, I don't expect that the cluster admin continues to switch on and off the feature gate as it happens in the tests.

vasiliy-ul commented:
> [One](https://github.com/kubevirt/kubevirt/blob/main/tests/operator/operator.go#L1599) of the functional tests in the operator series tries to install the image from the previous version, and of course we don't have the pr-helper image.

@alicefr, Why would the previous version try to use the `pr-helper` image? If I understand correctly, the test does the following: it installs `v0.58.1` (i.e. the previous release) and then tries to update it to the current version. Now, since in v0.58.1 the functionality introduced by this PR was not there, the corresponding `virt-handler` pod should not include the new container, right? Even if the feature gate is enabled, it should be ignored since v0.58.1 does not know anything about it. Or did I get it wrong?

> enabling the PersistentReservation triggers virt-handler to be redeployed and this slows the entire system down. 

Also wanted to clarify: you mention that only virt-handler is redeployed. In fact, according to my observation, virt-api and virt-controller are redeployed as well:

```
$ k edit kubevirt -n kubevirt kubevirt
kubevirt.kubevirt.io/kubevirt edited

$ k get pod -n kubevirt -w
NAME                               READY   STATUS    RESTARTS   AGE
disks-images-provider-4drql        1/1     Running   0          2m55s
virt-api-7db4795d9-dgt7w           1/1     Running   0          2m14s
virt-controller-7c745b44cd-k4wqj   1/1     Running   0          108s
virt-controller-7c745b44cd-mqh8s   1/1     Running   0          108s
virt-handler-xlvqq                 1/1     Running   0          108s
virt-operator-7656d6888c-f4nfz     1/1     Running   0          2m51s
virt-operator-7656d6888c-gm65z     1/1     Running   0          2m51s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Pending   0          0s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Pending   0          0s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     ContainerCreating   0          0s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     ContainerCreating   0          1s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   1/1     Running             0          2s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          6s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          7s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          8s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          8s
virt-handler-xlvqq                                                1/1     Terminating         0          2m35s
virt-handler-xlvqq                                                1/1     Terminating         0          2m36s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Terminating         0          11s
kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Terminating         0          11s
virt-handler-xlvqq                                                0/1     Terminating         0          2m36s
virt-handler-xlvqq                                                0/1     Terminating         0          2m36s
virt-handler-xlvqq                                                0/1     Terminating         0          2m36s
virt-handler-ldkhp                                                0/2     Pending             0          0s
virt-handler-ldkhp                                                0/2     Pending             0          0s
virt-handler-ldkhp                                                0/2     Init:0/1            0          0s
virt-handler-ldkhp                                                0/2     Init:0/1            0          1s
virt-handler-ldkhp                                                0/2     Init:0/1            0          1s
virt-handler-ldkhp                                                0/2     PodInitializing     0          2s
virt-handler-ldkhp                                                1/2     Running             0          6s
virt-handler-ldkhp                                                2/2     Running             0          20s
virt-controller-5dc54884f9-85pkj                                  0/1     Pending             0          0s
virt-controller-5dc54884f9-85pkj                                  0/1     Pending             0          0s
virt-controller-5dc54884f9-85pkj                                  0/1     ContainerCreating   0          0s
virt-controller-5dc54884f9-85pkj                                  0/1     ContainerCreating   0          0s
virt-controller-5dc54884f9-85pkj                                  0/1     Running             0          1s
virt-controller-5dc54884f9-85pkj                                  1/1     Running             0          20s
virt-controller-7c745b44cd-k4wqj                                  1/1     Terminating         0          3m22s
virt-controller-5dc54884f9-pwp9x                                  0/1     Pending             0          0s
virt-controller-5dc54884f9-pwp9x                                  0/1     Pending             0          0s
virt-controller-5dc54884f9-pwp9x                                  0/1     ContainerCreating   0          0s
virt-controller-7c745b44cd-k4wqj                                  1/1     Terminating         0          3m22s
virt-controller-5dc54884f9-pwp9x                                  0/1     ContainerCreating   0          1s
virt-controller-5dc54884f9-pwp9x                                  0/1     Running             0          1s
virt-controller-7c745b44cd-k4wqj                                  0/1     Terminating         0          3m23s
virt-controller-7c745b44cd-k4wqj                                  0/1     Terminating         0          3m23s
virt-controller-7c745b44cd-k4wqj                                  0/1     Terminating         0          3m23s
virt-controller-5dc54884f9-pwp9x                                  1/1     Running             0          20s
virt-controller-7c745b44cd-mqh8s                                  1/1     Terminating         0          3m43s
virt-controller-7c745b44cd-mqh8s                                  1/1     Terminating         0          3m43s
virt-controller-7c745b44cd-mqh8s                                  0/1     Terminating         0          3m43s
virt-controller-7c745b44cd-mqh8s                                  0/1     Terminating         0          3m43s
virt-controller-7c745b44cd-mqh8s                                  0/1     Terminating         0          3m43s
virt-api-56474fb4f6-mdt8l                                         0/1     Pending             0          0s
virt-api-56474fb4f6-mdt8l                                         0/1     Pending             0          0s
virt-api-56474fb4f6-mdt8l                                         0/1     ContainerCreating   0          0s
virt-api-56474fb4f6-mdt8l                                         0/1     ContainerCreating   0          0s
virt-api-56474fb4f6-mdt8l                                         0/1     Running             0          1s
virt-api-56474fb4f6-mdt8l                                         1/1     Running             0          20s
virt-api-7db4795d9-dgt7w                                          1/1     Terminating         0          4m34s
virt-api-7db4795d9-dgt7w                                          1/1     Terminating         0          4m39s
virt-api-7db4795d9-dgt7w                                          0/1     Terminating         0          4m40s
virt-api-7db4795d9-dgt7w                                          0/1     Terminating         0          4m40s
virt-api-7db4795d9-dgt7w                                          0/1     Terminating         0          4m40s
```

Is that the expected behavior?

alicefr commented:
> > [One](https://github.com/kubevirt/kubevirt/blob/main/tests/operator/operator.go#L1599) of the functional tests in the operator series tries to install the image from the previous version, and of course we don't have the pr-helper image.
> 
> @alicefr, Why would the previous version try to use the `pr-helper` image? If I understand correctly, the test does the following: it installs `v0.58.1` (i.e. the previous release) and then tries to update it to the current version. Now, since in v0.58.1 the functionality introduced by this PR was not there, the corresponding `virt-handler` pod should not include the new container, right? Even if the feature gate is enabled, it should be ignored since v0.58.1 does not know anything about it. Or did I get it wrong?

It is `virt-operator` that calls [`NewHandlerDaemonSet`](https://github.com/alicefr/kubevirt/blob/scsi-pr-device-plugin/pkg/virt-operator/resource/generate/install/strategy.go#L516). This isn't redeployed and it still has the information of the feature gate PersistentReservation and it enables the container

> 
> > enabling the PersistentReservation triggers virt-handler to be redeployed and this slows the entire system down.
> 
> Also wanted to clarify: you mention that only virt-handler is redeployed. In fact, according to my observation, virt-api and virt-controller are redeployed as well:
> 
> ```
> $ k edit kubevirt -n kubevirt kubevirt
> kubevirt.kubevirt.io/kubevirt edited
> 
> $ k get pod -n kubevirt -w
> NAME                               READY   STATUS    RESTARTS   AGE
> disks-images-provider-4drql        1/1     Running   0          2m55s
> virt-api-7db4795d9-dgt7w           1/1     Running   0          2m14s
> virt-controller-7c745b44cd-k4wqj   1/1     Running   0          108s
> virt-controller-7c745b44cd-mqh8s   1/1     Running   0          108s
> virt-handler-xlvqq                 1/1     Running   0          108s
> virt-operator-7656d6888c-f4nfz     1/1     Running   0          2m51s
> virt-operator-7656d6888c-gm65z     1/1     Running   0          2m51s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Pending   0          0s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Pending   0          0s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     ContainerCreating   0          0s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     ContainerCreating   0          1s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   1/1     Running             0          2s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          6s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          7s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          8s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Completed           0          8s
> virt-handler-xlvqq                                                1/1     Terminating         0          2m35s
> virt-handler-xlvqq                                                1/1     Terminating         0          2m36s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Terminating         0          11s
> kubevirt-2b9a2334d4ad4c4557776efb228c35e7d8a24d1a-jobgppmmljwmh   0/1     Terminating         0          11s
> virt-handler-xlvqq                                                0/1     Terminating         0          2m36s
> virt-handler-xlvqq                                                0/1     Terminating         0          2m36s
> virt-handler-xlvqq                                                0/1     Terminating         0          2m36s
> virt-handler-ldkhp                                                0/2     Pending             0          0s
> virt-handler-ldkhp                                                0/2     Pending             0          0s
> virt-handler-ldkhp                                                0/2     Init:0/1            0          0s
> virt-handler-ldkhp                                                0/2     Init:0/1            0          1s
> virt-handler-ldkhp                                                0/2     Init:0/1            0          1s
> virt-handler-ldkhp                                                0/2     PodInitializing     0          2s
> virt-handler-ldkhp                                                1/2     Running             0          6s
> virt-handler-ldkhp                                                2/2     Running             0          20s
> virt-controller-5dc54884f9-85pkj                                  0/1     Pending             0          0s
> virt-controller-5dc54884f9-85pkj                                  0/1     Pending             0          0s
> virt-controller-5dc54884f9-85pkj                                  0/1     ContainerCreating   0          0s
> virt-controller-5dc54884f9-85pkj                                  0/1     ContainerCreating   0          0s
> virt-controller-5dc54884f9-85pkj                                  0/1     Running             0          1s
> virt-controller-5dc54884f9-85pkj                                  1/1     Running             0          20s
> virt-controller-7c745b44cd-k4wqj                                  1/1     Terminating         0          3m22s
> virt-controller-5dc54884f9-pwp9x                                  0/1     Pending             0          0s
> virt-controller-5dc54884f9-pwp9x                                  0/1     Pending             0          0s
> virt-controller-5dc54884f9-pwp9x                                  0/1     ContainerCreating   0          0s
> virt-controller-7c745b44cd-k4wqj                                  1/1     Terminating         0          3m22s
> virt-controller-5dc54884f9-pwp9x                                  0/1     ContainerCreating   0          1s
> virt-controller-5dc54884f9-pwp9x                                  0/1     Running             0          1s
> virt-controller-7c745b44cd-k4wqj                                  0/1     Terminating         0          3m23s
> virt-controller-7c745b44cd-k4wqj                                  0/1     Terminating         0          3m23s
> virt-controller-7c745b44cd-k4wqj                                  0/1     Terminating         0          3m23s
> virt-controller-5dc54884f9-pwp9x                                  1/1     Running             0          20s
> virt-controller-7c745b44cd-mqh8s                                  1/1     Terminating         0          3m43s
> virt-controller-7c745b44cd-mqh8s                                  1/1     Terminating         0          3m43s
> virt-controller-7c745b44cd-mqh8s                                  0/1     Terminating         0          3m43s
> virt-controller-7c745b44cd-mqh8s                                  0/1     Terminating         0          3m43s
> virt-controller-7c745b44cd-mqh8s                                  0/1     Terminating         0          3m43s
> virt-api-56474fb4f6-mdt8l                                         0/1     Pending             0          0s
> virt-api-56474fb4f6-mdt8l                                         0/1     Pending             0          0s
> virt-api-56474fb4f6-mdt8l                                         0/1     ContainerCreating   0          0s
> virt-api-56474fb4f6-mdt8l                                         0/1     ContainerCreating   0          0s
> virt-api-56474fb4f6-mdt8l                                         0/1     Running             0          1s
> virt-api-56474fb4f6-mdt8l                                         1/1     Running             0          20s
> virt-api-7db4795d9-dgt7w                                          1/1     Terminating         0          4m34s
> virt-api-7db4795d9-dgt7w                                          1/1     Terminating         0          4m39s
> virt-api-7db4795d9-dgt7w                                          0/1     Terminating         0          4m40s
> virt-api-7db4795d9-dgt7w                                          0/1     Terminating         0          4m40s
> virt-api-7db4795d9-dgt7w                                          0/1     Terminating         0          4m40s
> ```
> 
> Is that the expected behavior?

I think yes, however I'm not very familiar with the downgrade process

alicefr commented:
After some discussions with @vladikr and @xpivarc and investigation, there are the following updates on the tests for this PR: 
  1. The `PersistentReservation` feature gate enablement/removal redeploys virt-handler and for this reason, the tests that use this feature gate need to run serially as restarting virt-handler can interfere with the other tests run in parallel
  2. The `PersistentReservation` feature gate won't be enabled by default for the tests. Certain tests modify the KubeVirt CR and this might cause flakiness in the already existing tests because of virt-handler restarts
  3. The operator test mentioned in the [comment](https://github.com/kubevirt/kubevirt/pull/9177#issuecomment-1448195122) above has a bug. The right path is that the operator is the same version as the other components. The bug will be investigated separately from this PR, and currently, we don't enable the `PersistentReservation` for this test.

@vasiliy-ul some of these points answer some of your reviews and feedback on this PR.

alicefr commented:
/retest-required

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.25-sig-operator

xpivarc reviewed (commented):
First pass. I did not catch anything critical

alicefr commented:
/retest-required

alicefr commented:
/retest-required

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.25-sig-storage

alicefr commented:
@vasiliy-ul @xpivarc @vladikr could you please have a second look to this PR when you have time?
I have addressed the review comments and done a couple of improvements.

vladikr commented on api/api-rule-violations-known.list:
```diff
@@ -280,6 +280,7 @@ API rule violation: names_match,kubevirt.io/api/core/v1,KubeVirtConfiguration,SE
 API rule violation: names_match,kubevirt.io/api/core/v1,KubeVirtConfiguration,SMBIOSConfig
 API rule violation: names_match,kubevirt.io/api/core/v1,KubeVirtSpec,CertificateRotationStrategy
 API rule violation: names_match,kubevirt.io/api/core/v1,LunTarget,ReadOnly
+API rule violation: names_match,kubevirt.io/api/core/v1,LunTarget,Reservation
```
I think you can easily avoid this violation, just make the field name and the json name match

alicefr commented on cmd/pr-helper/BUILD.bazel:
```diff
@@ -0,0 +1,43 @@
+pkg_tar(
+    name = "multipath",
+    srcs = [":multipath.conf"],
+    mode = "0300",
```
oh yes, good catch!

alicefr commented:
Rebased

alicefr commented:
Addressed the last review comments. Anything else to change?

vasiliy-ul reviewed (approved):
Did one more review round and no new issues found. Since there are no objections on implementation approach and the CI is happy, I think this is good to get merged.

alicefr commented:
@vasiliy-ul I asked @xpivarc to give it a second look

xpivarc commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -399,6 +400,11 @@ func (app *virtHandlerApp) Run() {
 	go gracefulShutdownInformer.Run(stop)
 	go domainSharedInformer.Run(stop)
 
+	prSockDir, err := safepath.JoinAndResolveWithRelativeRoot("/", reservation.GetPrHelperSocketDir())
```
(I am ok if you resolve this comment in follow up). 
I believe we don't need to do this here. I would move the relabeling into context of virt-launcher. Meaning, the virt-launcher is running and before first `sync` is send we can adjust the provided socket to the targeted SELinux context.

alicefr replied:
I did it here because we are labeling at this point also the devices: https://github.com/kubevirt/kubevirt/blob/main/cmd/virt-handler/virt-handler.go#L404-L419

xpivarc commented on pkg/virt-handler/device-manager/socket_device.go:
```diff
@@ -0,0 +1,326 @@
+	pluginapi "kubevirt.io/kubevirt/pkg/virt-handler/device-manager/deviceplugin/v1beta1"
+)
+
+type SocketDevicePlugin struct {
```
We should cover this by unit tests.

xpivarc commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -544,6 +540,35 @@ func (app *virtHandlerApp) shouldInstallKubevirtSeccompProfile() {
 
 }
 
+// Update virt-handler rate limiter
```
Comment is miss-leading.

alicefr replied:
Removed

xpivarc commented on tests/storage/reservation.go:
```diff
@@ -0,0 +1,295 @@
+		Expect(err).ToNot(HaveOccurred())
+
+		stdout, stderr, err := exec.ExecuteCommandOnPodWithResults(virtClient, pod, "targetcli", cmd)
+		By(fmt.Sprintf("targetcli: stdout: %v stderr: %v", stdout, stderr))
```
This is miss use of `By`. You probably want `Expect(err).ToNot(HaveOccurred(), stdout, stderr)`

xpivarc reviewed (approved):
/hold cancel
I am happy if we do follow ups after this.

alicefr commented:
/retest-required

alicefr commented:
@xpivarc if I need a rebase, I'll include them otherwise, I'll address in a second PR

xpivarc commented on tests/storage/reservation.go:
```diff
@@ -0,0 +1,295 @@
+	BeforeEach(func() {
+		fgDisabled = !checks.HasFeature(virtconfig.PersistentReservation)
+		if fgDisabled {
+			tests.EnableFeatureGate(virtconfig.PersistentReservation)
```
We  need to use "custom" `EnableFeatureGate` which will check the endpoint only after the re-deployment.

xpivarc reviewed (changes requested):
@alicefr I think we need to address the failing test. It is pretty consistent. The problem is that updating the feature gate is triggering the re-deployment of virt-handler. Therefore it fails when we check the endpoint.  I still think all the other points can be addressed in follow up to easy the merging.

alicefr commented:
> @alicefr I think we need to address the failing test. It is pretty consistent. The problem is that updating the feature gate is triggering the re-deployment of virt-handler. Therefore it fails when we check the endpoint. I still think all the other points can be addressed in follow up to easy the merging.

Yes, previously I had increased the time out for virt-handler: https://github.com/kubevirt/kubevirt/blob/main/tests/utils.go#L1821 and the failure was gone. I have removed that commit because of  https://github.com/kubevirt/kubevirt/commit/67bfbf47ea62b2b5242dae7395a1622a2cb8fbe3. But apparently we have moved back to 10 sec, my change was 20s

alicefr commented on tests/storage/reservation.go:
```diff
@@ -0,0 +1,295 @@
+			vmi := libvmi.NewFedora(
+				libvmi.WithPersistentVolumeClaimLun("lun0", pvc.Name, true),
+			)
+			vmi.Namespace = util.NamespaceTestDefault
```
The main reason I added, it is because the function [CreatePVandPVCwithSCSIDisk](https://github.com/kubevirt/kubevirt/blob/f9566fe4f85bb592b09faefb80b1406395043027/tests/storage/reservation.go#L208) has the namespace and I wanted to be sure they are created on the same namespace

alicefr commented:
Rebased and addressed ci failure in commit: https://github.com/kubevirt/kubevirt/pull/9177/commits/fccb13994921d3fc63dac32fe8805fe79661815a

xpivarc commented:
/lgtm

alicefr commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -544,6 +540,35 @@ func (app *virtHandlerApp) shouldInstallKubevirtSeccompProfile() {
+// Update virt-handler rate limiter
+func (app *virtHandlerApp) shouldEnablePersistentReservation() {
+	enabled := app.clusterConfig.PersistentReservationEnabled()
+	if !enabled {
```
@xpivarc do you mean moving the labeling on the device plugin activation?

xpivarc replied:
Here I believe I was referring to the plugin itself

alicefr replied:
mh it is already activated based on the feature gate in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/device-manager/device_controller.go#L193
